File: csharp/src/Training.Workshop/M7/S10BooleanParameter/Step1/TicketService.cs
using System.Globalization;

namespace Training.Workshop.M7.S10BooleanParameter.Step1;

/// <summary>
/// Krok 1: flaga online zastąpiona jawnymi metodami BookOnline / BookAtBoxOffice.
/// Stara metoda zostaje jako [Obsolete] i deleguje - klienci migrują po jednym.
/// Stara sygnatura zostaje jako osobne przeciążenie: skompilowany wcześniej klient z innego
/// assembly odwołuje się do niej w IL, więc jej usunięcie łamie zgodność binarną.
/// </summary>
public sealed class TicketService
{
    private const decimal GlassesPrice = 3.00m;
    private const decimal Fee = 2.00m;

    private enum Channel { Online, BoxOffice }

    /// <summary>Przestarzałe: użyj <see cref="BookOnline"/> albo <see cref="BookAtBoxOffice"/>.</summary>
    [Obsolete("użyj BookOnline albo BookAtBoxOffice")]
    public string Book(string title, string format, int seats, bool online, bool ownGlasses)
    {
        return online
            ? BookOnline(title, format, seats, ownGlasses)
            : BookAtBoxOffice(title, format, seats, ownGlasses);
    }

    public string BookOnline(string title, string format, int seats, bool ownGlasses)
    {
        return Book(title, format, seats, Channel.Online, ownGlasses);
    }

    public string BookAtBoxOffice(string title, string format, int seats, bool ownGlasses)
    {
        return Book(title, format, seats, Channel.BoxOffice, ownGlasses);
    }

    private string Book(string title, string format, int seats, Channel channel, bool ownGlasses)
    {
        var basePrice = format switch
        {
            "IMAX" => 40.00m,
            "3D" => 32.00m,
            _ => 25.00m,
        };
        decimal count = seats;
        var total = basePrice * count;
        if (format == "3D" && !ownGlasses)
        {
            total += GlassesPrice * count;
        }
        if (channel == Channel.Online)
        {
            total += Fee * count;
        }
        return title + " " + format + " x" + seats
            + (channel == Channel.Online ? " online" : " kasa") + ": "
            + total.ToString(CultureInfo.InvariantCulture);
    }
}

File: typescript/src/workshop/m7/s10_booleanparameter/step2/TicketService.ts
import { Decimal } from 'decimal.js';

import { Glasses } from './Glasses.js';

enum Channel { ONLINE = 'ONLINE', BOX_OFFICE = 'BOX_OFFICE' }

/**
 * Krok 2: druga flaga (ownGlasses) w nowych metodach zamieniona na enum Glasses.
 * Nie dokładamy metody na każdą kombinację (4 metody) - kanał to metoda, okulary to wartość.
 * Stara metoda @deprecated tłumaczy boolean na Glasses.
 */
export class TicketService {
  private static readonly GLASSES = new Decimal('3.00');
  private static readonly FEE = new Decimal('2.00');

  /**
   * @deprecated użyj {@link TicketService.bookOnline} albo {@link TicketService.bookAtBoxOffice}
   */
  book(title: string, format: string, seats: number, online: boolean, ownGlasses: boolean): string {
    const glasses = ownGlasses ? Glasses.OWN : Glasses.RENTED;
    return online
      ? this.bookOnline(title, format, seats, glasses)
      : this.bookAtBoxOffice(title, format, seats, glasses);
  }

  bookOnline(title: string, format: string, seats: number, glasses: Glasses): string {
    return this.bookInChannel(title, format, seats, Channel.ONLINE, glasses);
  }

  bookAtBoxOffice(title: string, format: string, seats: number, glasses: Glasses): string {
    return this.bookInChannel(title, format, seats, Channel.BOX_OFFICE, glasses);
  }

  // W Javie prywatne przeciążenie book(..., Channel, ...); TS nie ma przeciążeń - osobna nazwa.
  private bookInChannel(title: string, format: string, seats: number, channel: Channel, glasses: Glasses): string {
    let base: Decimal;
    switch (format) {
      case 'IMAX': base = new Decimal('40.00'); break;
      case '3D': base = new Decimal('32.00'); break;
      default: base = new Decimal('25.00');
    }
    const count = new Decimal(seats);
    let total = base.times(count);
    if (format === '3D' && glasses === Glasses.RENTED) {
      total = total.plus(TicketService.GLASSES.times(count));
    }
    if (channel === Channel.ONLINE) {
      total = total.plus(TicketService.FEE.times(count));
    }
    return `${title} ${format} x${seats}${channel === Channel.ONLINE ? ' online' : ' kasa'}: ${total.toFixed(2)}`;
  }
}
